README.md
# Cordova Hot Code Push Local Development Add-on

This plugin adds local development feature to the [Hot Code Push plugin](https://github.com/nordnet/cordova-hot-code-push). With it you can develop your app and see the result immediately on the device. No need to restart it on every change you make in your web files: all the updates are uploaded automatically.

## Supported platforms
- Android 4.0.0 or above.
- iOS 8.0 or above. Requires Xcode 8.

## Documentation
- [Installation](#installation)
- [Quick start guide for Cordova project](#quick-start-guide-for-cordova-project)
- [Quick start guide for Ionic project](#quick-start-guide-for-ionic-project)
- [Configuration](#configuration)
- [How it works](#how-it-works)
- [When to use this plugin](#when-to-use-this-plugin)
- [How to uninstall](#how-to-uninstall)

### Installation

This requires cordova 5.0+ (current stable 0.4.2):

```sh
cordova plugin add cordova-hot-code-push-local-dev-addon
```

It is also possible to install via repo url directly (__unstable__):
```sh
cordova plugin add https://github.com/nordnet/cordova-hot-code-push-local-dev-addon.git
```

Starting from `v0.4.0` this plugin **requires Xcode 8**. If you are using Xcode 7 - please, use `v0.3.0`:
```sh
cordova plugin add cordova-hot-code-push-local-dev-addon@0.3.0
```

**Note:** you can install plugin to any project, but it is not gonna do anything without the [Hot Code Push plugin](https://github.com/nordnet/cordova-hot-code-push).

### Quick start guide for Cordova project

1. Create new Cordova project using command line interface and add iOS/Android platforms:

  ```sh
  cordova create TestProject com.example.testproject TestProject
  cd ./TestProject
  cordova platform add android
  cordova platform add ios
  ```
  Or use the existing one.

2. Add Hot Code Push plugin:

  ```sh
  cordova plugin add cordova-hot-code-push-plugin
  ```

3. Add local development feature:

  ```sh
  cordova plugin add cordova-hot-code-push-local-dev-addon
  ```

4. Install Cordova Hot Code Push CLI client:

  ```sh
  npm install -g cordova-hot-code-push-cli
  ```

5. Start local server by executing:

  ```sh
  cordova-hcp server
  ```

  As a result you will see something like this:
  ```
  Running server
  Checking:  /Cordova/TestProject/www
  local_url http://localhost:31284
  Warning: .chcpignore does not exist.
  Build 2015.09.02-10.17.48 created in /Cordova/TestProject/www
  cordova-hcp local server available at: http://localhost:31284
  cordova-hcp public server available at: https://5027caf9.ngrok.com
  ```

6. Open new console window, go to the project root and launch the app:

  ```sh
  cordova run
  ```

  Wait until application is launched for both platforms.

6. Now open your `index.html` page in `www` folder of the `TestProject`, change something in it and save. In a few seconds you will see updated page on the launched devices (emulators).

From this point you can do local development, where all the changes are uploaded on the devices without the need to restart applications on every change you made.

### Quick start guide for Ionic project

1. Create new Ionic project using command line interface and add iOS/Android platforms:

  ```sh
  ionic start TestProject blank
  cd ./TestProject
  ionic platform add android
  ionic platform add ios
  ```
  Or use the existing one.

2. Build project for the first time before adding plugin to it.

  ```sh
  ionic build
  ```

  This is required mainly for iOS, because in some cases Ionic creates iOS project with the wrong name (`HelloCordova`) instead the one that is specified in the `config.xml`. But after the `build` it becomes the correct one.

3. Add Hot Code Push plugin:

  ```sh
  ionic plugin add cordova-hot-code-push-plugin
  ```

4. Add local development feature:

  ```sh
  ionic plugin add cordova-hot-code-push-local-dev-addon
  ```

5. Install Cordova Hot Code Push CLI client:

  ```sh
  npm install -g cordova-hot-code-push-cli
  ```

6. Start local server by executing:

  ```sh
  cordova-hcp server
  ```

  As a result you will see something like this:
  ```
  Running server
  Checking:  /Cordova/TestProject/www
  local_url http://localhost:31284
  Warning: .chcpignore does not exist.
  Build 2015.09.02-10.17.48 created in /Cordova/TestProject/www
  cordova-hcp local server available at: http://localhost:31284
  cordova-hcp public server available at: https://5027caf9.ngrok.com
  ```

7. Open new console window, go to the project root and launch the app:

  ```sh
  ionic run
  ```

  Wait until application is launched for both platforms.

8. Now open your `index.html` page in `www` folder of the `TestProject`, change something in it and save. In a few seconds you will see updated page on the launched devices (emulators).

From this point you can do local development, where all the changes are uploaded on the devices without the need to restart applications on every change you made.

### Configuration

If you just started a new project or didn't define `config-file` in the `config.xml` - then you don't need to do any configuration. You just need to launch the local server by executing `cordova-hcp server`, and the plugin will handle the rest.

But if your `config.xml` already has a `config-file` - then you need to explicitly activate plugin in the app like so:
```xml
<chcp>
  <config-file url="http://mydomain.com/chcp.json" />
  <local-development enabled="true" />
</chcp>
```

As you can see, you just need to add `<local-development />` preference to the `<chcp />` block and set `enabled` to `true`. If it set to `false` - plugin is disabled.

If enabled and local server is running - plugin will take it's url and inject it as a `config-file` in the platform-specific `config.xml`. As a result, your app will be configured to work with the local server.

### How it works

Plugin does his work on two stages:

1. When the project is build (`cordova build` or `cordova run`).
2. When the app is running.

##### Build phase

When you execute `cordova build` plugin:

1. Takes local server's url and inject's it as a `config-file` into platform-specific `config.xml`.
2. Increases build version of the app, to force reinstallation of the of the `www` folder on the external storage. Basically, this reset's the application's cache folder.
3. For iOS app it activates support for Swift code.

As an output you can see something like this in the console:
```
CHCP Local Development Add-on:
    Setting config-file to local server: https://20a51a38.ngrok.com/chcp.json
    Android version code is set to 5
    iOS bundle version set to 4
Swift support activation hook:
    IOS project now has deployment target set to: 7.0
    IOS project option EMBEDDED_CONTENT_CONTAINS_SWIFT set as: YES
    IOS project Runpath Search Paths set to: @executable_path/Frameworks
    Injected swift header HelloCordova-Swift.h into plugin's main header.
```

##### App is running

When the app launches - plugin tries to connect to the local server's socket. Thru that socket we will receive notifications from the server, that new release is ready for upload:
```
File changed:  /Cordova/Test/www/index.html
Build 2015.11.10-10.14.00 created in /Cordova/Test/www
Should trigger reload for build: 2015.11.10-10.14.00
```
When this happens - plugin captures that event and sends command to the main plugin, that it need to fetch the update.

As long as you work on your web project - there is no need to restart the app. The only case when you need to restart it is when you add new plugins. They can have some additional hooks, or native code, and we can't update that on-the-fly.

### When to use this plugin

You should use this plugin for development purpose only. It will help you to speed up the development, and nothing more. When the app is ready and you are prepared for release - don't forget to delete it. It's not gonna work in the release mode anyway. Plus, if you'll try to publish iOS version of the app with this plugin - there is a good chance that it will be rejected because of the Swift code. For some reason when you build your app from console - in the result archive there is no `SwiftSupport` folder. And because of that you will be rejected with the message:
```
Dear developer,

We have discovered one or more issues with your recent delivery for "Directory.". To process your delivery, the following issues must be corrected:

Invalid Swift Support - The SwiftSupport folder is missing. Rebuild your app using the current public (GM) version of Xcode and resubmit it.

Once these issues have been corrected, you can then redeliver the corrected binary.

Regards,

The App Store team
```

### How to uninstall

To remove the plugin just execute:
```sh
cordova plugin remove cordova-hot-code-push-local-dev-addon
```

As a result, it will delete all plugin's files. In the case of the iOS it will also disable Swift support in the project's config by setting `Embedded Content Contains Swift Code` preference to `NO`:
```
CHCP Local Dev plugin cleanup:
    IOS project option EMBEDDED_CONTENT_CONTAINS_SWIFT set as: NO
```


CHANGELOG.md
# Change Log

## v0.4.2 (2016-10-04)

**Bug fixes:**

- [Issue #14](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/14). Changed library to parse `.plist`. Previous one didn't work on some operating systems.

## v0.4.1 (2016-09-30)

**Bug fixes:**

- [Issue #14](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/14). Fixed corruption of project's `.plist` file when used with some other plugins (i.e. `cordova-plugin-camera`).

## v0.4.0 (2016-09-21)

**Bug fixes:**

- [Issue #13](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/13). Project can now be build with Xcode 8. To run in Xcode 7 use v0.3.0.

## v0.3.0 (2016-08-29)

**Bug fixes:**

- Fixed [issue #12](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/12). Problem was with how `plist` is parsed and saved.

**Enhancements:**

- Updated socket.io library for iOS to v7.0.3.

## v0.2.2 (2016-04-27)

**Bug fixes:**

- Fixed installation failure on Windows.

## v0.2.1 (2016-04-25)

**Bug fixes:**

- Fixed [issue #7](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/7). `<native-interface />` property is now preserved.
- Fixed typo in `before_plugin_install` hook. If installation has failed - it will show the correct error message.

## v0.2.0 (2016-04-22)

**Enhancements:**

- Dependency node modules are now installed in the plugins folder instead of the projects root folder.
- Updated Socket.IO library for iOS. Now it's version is 6.1.0.
- Updated Socket.IO library for Android. Now it's version is 0.7.

## v0.1.2 (2015-12-04)

**Bug fixes:**

- Fixed [issue #4](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/4).

## v0.1.1 (2015-10-14)

**Bug fixes:**

- Fixed [issue #2](https://github.com/nordnet/cordova-hot-code-push-local-dev-addon/issues/2).

**Docs:**

- Added change log file.


scripts/lib/swiftHelper.js
/*
Helper class to work with Swift.
Mainly, it has only two method: to activate and to deactivate swift support in the project.
*/

var path = require('path');
var fs = require('fs');
var logger = require('./logger.js');
var IOS_DEPLOYMENT_TARGET = '8.0';
var COMMENT_KEY = /_comment$/;
var PLUGIN_NAME = 'cordova-hot-code-push-local-dev-addon';
var PLUGIN_MAIN_HEADER = 'HCPLDPlugin.h';
var context;
var projectRoot;
var projectName;
var projectModuleName;
var iosPlatformPath;

module.exports = {
  activate: activate,
  disable: disable
};

/**
 * Activate swift support in the project.
 *
 * @param {Object} cordovaContext - cordova context
 */
function activate(cordovaContext) {

  init(cordovaContext);

  // injecting options in project file
  var projectFile = loadProjectFile();
  injectSwiftOptionsInProjectConfig(projectFile.xcode);
  projectFile.write();

  // injecting inclusion of Swift header
  injectSwiftHeader();
}

/**
 * Remove swift support from the project.
 *
 * @param {Object} cordovaContext - cordova context
 */
function disable(cordovaContext) {
  init(cordovaContext);

  var projectFile = loadProjectFile();
  removeSwiftOptionsFromProjectConfig(projectFile.xcode);
  projectFile.write();
}

// region General private methods

/**
 * Initialize before execution.
 *
 * @param {Object} ctx - cordova context instance
 */
function init(ctx) {
  context = ctx;
  projectRoot = ctx.opts.projectRoot;
  projectName = getProjectName(ctx, projectRoot);
  iosPlatformPath = path.join(projectRoot, 'platforms', 'ios')
}

/**
 * Load iOS project file from platform specific folder.
 *
 * @return {Object} projectFile - project file information
 */
function loadProjectFile() {
  var platform_ios,
    projectFile;

  try {
    // try pre-5.0 cordova structure
    platform_ios = context.requireCordovaModule('cordova-lib/src/plugman/platforms')['ios'];
    projectFile = platform_ios.parseProjectFile(iosPlatformPath);
  } catch (e) {
    // let's try cordova 5.0 structure
    platform_ios = context.requireCordovaModule('cordova-lib/src/plugman/platforms/ios');
    projectFile = platform_ios.parseProjectFile(iosPlatformPath);
  }

  return projectFile;
}

/**
 * Get name of the current project.
 *
 * @param {Object} ctx - cordova context instance
 * @param {String} projectRoot - current root of the project
 *
 * @return {String} name of the project
 */
function getProjectName(ctx, projectRoot) {
  var cordova_util = ctx.requireCordovaModule('cordova-lib/src/cordova/util'),
    xml = cordova_util.projectConfig(projectRoot),
    ConfigParser;

  // If we are running Cordova 5.4 or abova - use parser from cordova-common.
  // Otherwise - from cordova-lib.
  try {
    ConfigParser = ctx.requireCordovaModule('cordova-common/src/ConfigParser/ConfigParser');
  } catch (e) {
    ConfigParser = ctx.requireCordovaModule('cordova-lib/src/configparser/ConfigParser')
  }

  return new ConfigParser(xml).name();
}

/**
 * Remove comments from the file.
 *
 * @param {Object} obj - file object
 * @return {Object} file object without comments
 */
function nonComments(obj) {
  var keys = Object.keys(obj),
    newObj = {};

  for (var i = 0, len = keys.length; i < len; i++) {
    if (!COMMENT_KEY.test(keys[i])) {
      newObj[keys[i]] = obj[keys[i]];
    }
  }

  return newObj;
}

// endregion

// region Swift options injection

/**
 * Inject Swift options into project configuration file.
 *
 * @param {Object} xcodeProject - xcode project file instance
 */
function injectSwiftOptionsInProjectConfig(xcodeProject) {
  var configurations = nonComments(xcodeProject.pbxXCBuildConfigurationSection()),
    config,
    buildSettings;

  for (config in configurations) {
    buildSettings = configurations[config].buildSettings;
    buildSettings['IPHONEOS_DEPLOYMENT_TARGET'] = IOS_DEPLOYMENT_TARGET;
    buildSettings['EMBEDDED_CONTENT_CONTAINS_SWIFT'] = "YES";
    buildSettings['LD_RUNPATH_SEARCH_PATHS'] = '"@executable_path/Frameworks"';
    buildSettings['SWIFT_VERSION'] = '3.0';

    // if project module name is not defined - set it with value from build settings
    if ((!projectModuleName || projectModuleName.length == 0) && buildSettings['PRODUCT_NAME']) {
      setProjectModuleName(buildSettings['PRODUCT_NAME']);
    }
  }
  logger.info('iOS project now has deployment target set to: ' + IOS_DEPLOYMENT_TARGET);
  logger.info('iOS project option EMBEDDED_CONTENT_CONTAINS_SWIFT set as: YES');
  logger.info('iOS project Runpath Search Paths set to: @executable_path/Frameworks');
  logger.info('iOS project "Use Legacy Swift Language Version" set to: NO');
}

/**
 * Set project module name from the build settings.
 * Will be used to generate name of the Swift header.
 *
 * @param {String} nameFromBuildSettings - name of the product from build settings
 */
function setProjectModuleName(nameFromBuildSettings) {
  projectModuleName = nameFromBuildSettings.trim().replace(/"/g, '');
}

/**
 * Inject Swift inclusion header into ProjectName-Prefix.pch.
 * This way we ensure that Swift libraries are accessible in all project classes.
 */
function injectSwiftHeader() {
  // path to Prefix file and the name of included header
  var pluginHeaderFilePath = path.join(iosPlatformPath, projectName, 'Plugins', PLUGIN_NAME, PLUGIN_MAIN_HEADER),
    swiftImportHeader = generateSwiftHeaderFromProjectName(projectModuleName),
    headerFileContent;

  // read header file
  try {
    headerFileContent = fs.readFileSync(pluginHeaderFilePath, 'utf8');
  } catch (err) {
    logger.error(err);
    return;
  }

  // don't import if it is already there
  if (headerFileContent.indexOf(swiftImportHeader) > -1) {
    return;
  }

  // set header
  headerFileContent = setSwiftHeader(headerFileContent, swiftImportHeader);

  // save changes
  fs.writeFileSync(pluginHeaderFilePath, headerFileContent, 'utf8');

  logger.info('Injected swift header ' + swiftImportHeader + ' into plugin\'s main header.');
}

/**
 * Set proper header for Swift support.
 *
 * @param {String} prefixFileContent - Prefix.pch file content
 * @param {String} swiftImportHeader - header to set
 * @return {String} Prefix.pch file content with correct Swift header import
 */
function setSwiftHeader(prefixFileContent, swiftImportHeader) {
  // remove old import if there is any;
  // can occur if we change product name
  var found = prefixFileContent.match(/(\"[a-z0-9_]+-Swift.h\")/i);
  if (found && found.length > 0) {
    prefixFileContent = prefixFileContent.replace(found[0], '"' + swiftImportHeader + '"');
  } else {
    prefixFileContent = '#import "' + swiftImportHeader + '"\n' + prefixFileContent;
  }

  return prefixFileContent;
}

/**
 * Generate name of the header file for Swift support.
 * Details on Swift header name could be found here: https://developer.apple.com/library/ios/documentation/Swift/Conceptual/BuildingCocoaApps/MixandMatch.html
 *
 * @param {String} projectModuleName - projects module name from build configuration
 * @return {String} Swift header name
 */
function generateSwiftHeaderFromProjectName(projectModuleName) {
  var normalizedName = projectModuleName.replace(/([^a-z0-9]+)/gi, '_');

  return normalizedName + '-Swift.h'
}

// endregion

// region Disable

/**
 * Remove Swift options from the project file.
 *
 * @param {Object} xcodeProject - xcode project
 */
function removeSwiftOptionsFromProjectConfig(xcodeProject) {
  var configurations = nonComments(xcodeProject.pbxXCBuildConfigurationSection()),
    config,
    buildSettings;

  for (config in configurations) {
    buildSettings = configurations[config].buildSettings;
    buildSettings['EMBEDDED_CONTENT_CONTAINS_SWIFT'] = "NO";
  }

  logger.info('IOS project option EMBEDDED_CONTENT_CONTAINS_SWIFT set as: NO');
}

// endregion
